lib: nrf_modem: trace: measure backend bitrate

MirkoCovizzi opened the pull request:
This commit adds measurement and logging of the trace backend
bitrate. Useful for measuring performance of trace backends.

Signed-off-by: Mirko Covizzi <mirko.covizzi@nordicsemi.no>

Chosen `bitrate` and `bps` naming convention because the name is shorter compared to using `throughput` everywhere. Also, bitrate is a standard unit of measure of throughput.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-8702
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

eivindj-nordic commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,54 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+static int num_bps;
+
+#if CONFIG_NRF_MODEM_LIB_TRACE_LOG_BITRATE
+#define LOG_PERIOD_MSEC K_MSEC(CONFIG_NRF_MODEM_LIB_TRACE_LOG_BITRATE_PERIOD_MSEC)
```
This is no longer msec.
```suggestion
#define LOG_BITRATE_PERIOD K_MSEC(CONFIG_NRF_MODEM_LIB_TRACE_LOG_BITRATE_PERIOD_MSEC)
```

MirkoCovizzi replied:
I don't understand this comment. Can you elaborate?

eivindj-nordic replied:
Discussed offline, but for reference, it was just a suggestion to change the name of the define as it is not a value in milliseconds.

eivindj-nordic commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -68,6 +124,13 @@ static int trace_fragment_write(struct nrf_modem_trace_data *frag)
 			LOG_WRN("trace_backend_write wrote 0 bytes.");
 		}
 
+		/* Prevent division by zero (sometimes `end` == `start`) */
+		if (end != start) {
+			bps = ret / (end - start) * 1000 * 8;
+
+			update_bps(bps);
+		}
+
 		remaining -= ret;
```
Happy path?
```suggestion
		remaining -= ret;
		if (end == start) {
		        /* Prevent division by zero (sometimes `end` == `start`) */
		        continue;
		}
		bps = ret / (end - start) * 1000 * 8;

		update_bps(bps);
```

MirkoCovizzi replied:
I don't like the idea of using `continue` like that (more confusing), but I will try to find a different way to write the code if this implementation looks unsatisfying.

eivindj-nordic reviewed (commented):
I was wondering if there should be a kconfig to enable the extra calculations as well, not just the logging. The bps calculations adds extra complexity (and execution time) to the trace library. Although this is in the area of 10s of us per trace fragment, while it takes ~90us - 80ms to process each fragment, it will affect the total trace throughput slightly.

MirkoCovizzi commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,54 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+static double min_bps = DOUBLE_MAX;
+static double max_bps;
+static double tot_bps;
+static int num_bps;
```
> I was wondering if there should be a kconfig to enable the extra calculations as well, not just the logging. The bps calculations adds extra complexity (and execution time) to the trace library. Although this is in the area of 10s of us per trace fragment, while it takes ~90us - 80ms to process each fragment, it will affect the total trace throughput slightly.

I thought about it and it would make the code much more verbose to little to no measurable improvement in performance. For example, the API function would only be available after toggling the option, which could create confusion. I'll see if I can find a better way to organize the code and please let me know if having a conditional API is a satisfying option and why.

NordicBuilder commented:
### Integration test specification

| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-fw-nrfconnect-nrf-iot_lwm2m | X | | |
| test-fw-nrfconnect-nrf-iot_mosh | X | | |
| test-fw-nrfconnect-nrf-iot_samples | X | | |
| test-fw-nrfconnect-nrf-iot_zephyr_lwm2m | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/26801/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

lemrey commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +37,92 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+	}
+}
+#else
+int nrf_modem_lib_trace_backend_bitrate_get(double *min_val, double *max_val, double *avg_val)
+{
+	return -EOPNOTSUPP;
+}
```
I would rather make the function not available when the option is turned off.
That way if this function is called the compiler will emit an error during compilation. That's better than adding runtime checks (code) to ignore this value in this specific case instead.

MirkoCovizzi replied:
I don't agree because if I were to use this function in an application, I would prefer to not worry about more conditional compilation just to compile successfully. I don't think having a conditional API (and `ifdef`s inside the header) is a good idea, but I'm going to look into Zephyr if there are existing cases like this, in which case I will re-evaluate.

MirkoCovizzi commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +37,92 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+
+static void log_bitrate_stats(struct k_work *item)
+{
+	LOG_INF("Trace bitrate (bps): %10.2lf", tot_bps / num_bps);
```
I tried on a fresh application (`download`) and it works by just enabling the feature via kconfig.

divipillai commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -655,6 +655,12 @@ Modem libraries
 
+    * Added:
+
+      * The kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BITRATE` option to enable the :c:func:`nrf_modem_lib_trace_bitrate_get` function that retrieves modem trace backend bitrate statistics.
+      * The kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BITRATE_LOG` option to enable logging of the modem trace backend bitrate.
+      * The kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BITRATE_LOG_PERIOD_MS` to set the period value of the modem trace backend bitrate logging.
```
```suggestion
      * The :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BITRATE` Kconfig option to enable the :c:func:`nrf_modem_lib_trace_bitrate_get` function that retrieves modem trace backend bitrate statistics.
      * The :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BITRATE_LOG` Kconfig option to enable logging of the modem trace backend bitrate.
      * The :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BITRATE_LOG_PERIOD_MS` Kconfig option to set the period value of the modem trace backend bitrate logging.
```
The `CONFIG_NRF_MODEM_LIB_TRACE_BITRATE_LOG_PERIOD_MS` Kconfig option, like the other Kconfig options, could be mentioned in the [nrf_modem_lib.rst](https://github.com/nrfconnect/sdk-nrf/pull/8702/files#diff-cbb37da4ceb2d24e21bdad1441e3c33e5bfad3377b5bcf23ab0a9f315d933110).

MirkoCovizzi replied:
It is not strictly required to mention that Kconfig option in the documentation (as for other similar time-related Kconfig options in the library).

lemrey commented on lib/nrf_modem_lib/Kconfig.modemlib:
```diff
@@ -77,6 +77,27 @@ config NRF_MODEM_LIB_TRACE_THREAD_PRIO
 	default 0
 
+config NRF_MODEM_LIB_TRACE_BITRATE
+	bool "Enable trace backend bitrate measurement"
```
What about "Measure trace backend bitrate"

MirkoCovizzi replied:
I prefer this because it mirrors the description for the tracing Kconfig option.

lemrey replied:
I think we should avoid "Enable" in the KConfig text since it appears next to a checkbox, and so it is implied that checking the box enables the feature.

lemrey commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +39,89 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+int nrf_modem_lib_trace_backend_bitrate_get(double *min_val, double *max_val, double *avg_val)
+{
+	if (min_val == NULL || max_val == NULL || avg_val == NULL) {
+		return -EINVAL;
```
Let's return `-EFAULT` in case of bad pointers

lemrey commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +39,89 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+
+	if (size > 0) {
+		/* Casting `int64_t` to `double` */
+		delta = bitrate_end - bitrate_start;
+		time = delta / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+		bps = size * 8 / time;
```
I think this would be simplified by using `k_uptime_get`/`k_uptime_get_32`, give it a look

MirkoCovizzi replied:
Not precise enough.

eivindj-nordic commented on include/modem/nrf_modem_lib_trace.h:
```diff
@@ -51,6 +51,18 @@ int nrf_modem_lib_trace_processing_done_wait(k_timeout_t timeout);
  */
 int nrf_modem_lib_trace_level_set(enum nrf_modem_lib_trace_level trace_level);
 
+#if CONFIG_NRF_MODEM_LIB_TRACE_BITRATE
```
If we want to show this in the documentation: 
```suggestion
#if CONFIG_NRF_MODEM_LIB_TRACE_BITRATE  || defined(__DOXYGEN__)
```

lemrey commented on include/modem/nrf_modem_lib_trace.h:
```diff
@@ -51,6 +51,20 @@ int nrf_modem_lib_trace_processing_done_wait(k_timeout_t timeout);
+ *
+ * @return Zero on success, non-zero otherwise.
+ */
+int nrf_modem_lib_trace_bitrate_get(double *value);
```
Why is the `bitrate` a double? This should probably be an integer.

MirkoCovizzi replied:
Because it's an average

lemrey replied:
Yeah, but it's an average over a number of bits.
There isn't anything smaller than a bit, so why are we interested in measuring `2.21908349832` bits instead `2` ?

eivindj-nordic replied:
I think having this as a double would make sense if the transfer rate was measured in e.g. `kbps` (kilo bits per second). As this is the rate in bits per second, and we are talking about quite high numbers in general (up to ~100 000 bps) I agree with @lemrey that a resolution of whole bits per second is high enough.

MirkoCovizzi replied:
Ok, I removed floating point completely.

lemrey reviewed (commented):
Did you consider adding a way to measure the throughput of the trace from the modem itself, in addition to the throughput of the trace backend? Note these are two different things:

- backend throughput: how many bits per second can the be written by the backend
- trace throughput: how many bits per second of trace data does the modem generate

MirkoCovizzi commented:
> Did you consider adding a way to measure the throughput of the trace from the modem itself, in addition to the throughput of the trace backend? Note these are two different things:
> 
>     * backend throughput: how many bits per second can the be written by the backend
> 
>     * trace throughput: how many bits per second of trace data does the modem generate

Yes, I've considered, but first I wanted to re-organize this one. I refactored and cleaned up a bit, before there were nested directives, now it's a bit cleaner. Also, I added a second period Kconfig, so we have one for sampling and one for logging, separate.

divipillai commented:
Do we need to add a change log entry for the PR?

MirkoCovizzi commented:
> Do we need to add a change log entry for the PR?

Yes

lemrey commented:
Sometimes the trace bps is negative
`[00:00:09.624,969] <inf> nrf_modem_lib_trace: Trace bitrate (bps): -22528742`

lemrey commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,182 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+
+static void backend_bps_log(struct k_work *item)
+{
+	LOG_INF("Trace backend bitrate (bps): %d", backend_bps_avg);
```
should this be `%u`?

lemrey commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,182 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
 static int trace_deinit(void);
 
+#if CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE
+static int backend_bps_avg;
+static int backend_bps_tot;
```
These should probably be `uint32_t`

lemrey commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,182 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+#endif
+
+#if CONFIG_NRF_MODEM_LIB_TRACE_BITRATE
+static int bps_avg;
+static int bps_tot;
```
`uint32_t`

lemrey commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,182 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+
+static void bps_log(struct k_work *item)
+{
+	LOG_INF("Trace bitrate (bps): %d", bps_avg);
```
and here, `%u`; if `%u` throws a warning, you might need `%lu` or `%llu`.

MirkoCovizzi commented:
> 

Could be useful to know the application that produced this output, thanks.

MirkoCovizzi commented:
> Sometimes the trace bps is negative `[00:00:09.624,969] <inf> nrf_modem_lib_trace: Trace bitrate (bps): -22528742`

What was the application that produced this output? thanks.

MirkoCovizzi commented:
> I set the measurement interval to 300

In which application?

lemrey commented:
oh, it was `samples/nrf9160/download`

eivindj-nordic commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,94 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+	backend_bps_samples++;
+}
+
+static void backend_bps_avg_update(struct k_work *item);
+
+K_WORK_DELAYABLE_DEFINE(backend_bps_avg_update_work, backend_bps_avg_update);
+
+static void backend_bps_avg_update(struct k_work *item)
+{
+	if (backend_bps_samples != 0) {
+		backend_bps_avg = backend_bps_tot / backend_bps_samples;
+	} else {
+		/* With 0 samples the average is 0 */
+		backend_bps_avg = 0;
+	}
+
+	backend_bps_reset();
+
+	k_work_schedule(&backend_bps_avg_update_work, BACKEND_BPS_AVG_UPDATE_PERIOD);
+}
```
Right... Thanks for clearifying :)

MirkoCovizzi commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,130 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+	}
+}
+#else
+static void trace_backend_bitrate_perf_start(void)
```
Ok, wrapped into macro so now we are sure that it's never going to appear if not enabled, and also we don't have empty functions with a comment inside.

eivindj-nordic commented on doc/nrf/libraries/modem/nrf_modem_lib.rst:
```diff
@@ -106,6 +106,12 @@ Passing ``NRF_MODEM_LIB_TRACE_LEVEL_OFF`` to the :c:func:`nrf_modem_lib_trace_le
 The application can synchronize with the flushing of modem traces by calling the :c:func:`nrf_modem_lib_trace_processing_done_wait` function.
 
+To enable the measurement of the modem trace backend bitrate, set the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE` Kconfig option to ``y`` in your project configuration.
+After enabling this Kconfig option, the application can use the :c:func:`nrf_modem_lib_trace_backend_bitrate_get` function to retrieve the rolling average bitrate of the modem trace backend, measured over the period defined by the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_PERIOD_MS` Kconfig option.
```
Maybe, add a sentence about how to select the `CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_PERIOD_MS`. If it is too short, the user will get 0 as the bitrate, even while traces are being processed.

eivindj-nordic commented on include/modem/nrf_modem_lib_trace.h:
```diff
@@ -51,6 +51,18 @@ int nrf_modem_lib_trace_processing_done_wait(k_timeout_t timeout);
 int nrf_modem_lib_trace_level_set(enum nrf_modem_lib_trace_level trace_level);
 
+#if defined(CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE) || defined(__DOXYGEN__)
+/** @brief Get the last measured average bitrate of the trace backend.
```
```suggestion
/** @brief Get the last measured rolling average bitrate of the trace backend.
```

eivindj-nordic commented on include/modem/nrf_modem_lib_trace.h:
```diff
@@ -51,6 +51,18 @@ int nrf_modem_lib_trace_processing_done_wait(k_timeout_t timeout);
+#if defined(CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE) || defined(__DOXYGEN__)
+/** @brief Get the last measured average bitrate of the trace backend.
+ *
+ * This function returns the last measured average bitrate of the trace
```
```suggestion
 * This function returns the last measured rolling average bitrate of the trace
```

eivindj-nordic commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -36,6 +36,124 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
+
+static uint32_t trace_bytes_received;
+
+static void update_trace_bytes_received(struct nrf_modem_trace_data *frags,
+					size_t n_frags)
```
```suggestion
static void update_trace_bytes_received(struct nrf_modem_trace_data *frags, size_t n_frags)
```

eivindj-nordic commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -85,10 +208,15 @@ void trace_thread_handler(void)
+
 		switch (err) {
-		case 0:
+		case 0: {
 			/* Success */
+
+			UPDATE_TRACE_BYTES_RECEIVED(frags, n_frags);
+
 			break;
+		}
```
```suggestion
		case 0:
			/* Success */
			UPDATE_TRACE_BYTES_RECEIVED(frags, n_frags);
			break;
```

eivindj-nordic reviewed (commented):
Looks good in general, so I did some nitpicking :)

eivindj-nordic commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -85,10 +207,13 @@ void trace_thread_handler(void)
+
 		switch (err) {
-		case 0:
+		case 0: {
```
```suggestion
		case 0:
```

eivindj-nordic commented on lib/nrf_modem_lib/nrf_modem_lib_trace.c:
```diff
@@ -85,10 +207,13 @@ void trace_thread_handler(void)
 			/* Success */
+			UPDATE_TRACE_BYTES_RECEIVED(frags, n_frags);
 			break;
+		}
```
```suggestion
		
```

divipillai commented on doc/nrf/libraries/modem/nrf_modem_lib.rst:
```diff
@@ -106,6 +106,13 @@ Passing ``NRF_MODEM_LIB_TRACE_LEVEL_OFF`` to the :c:func:`nrf_modem_lib_trace_le
 
+To enable the measurement of the modem trace backend bitrate, set the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE` Kconfig option to ``y`` in your project configuration.
+After enabling this Kconfig option, the application can use the :c:func:`nrf_modem_lib_trace_backend_bitrate_get` function to retrieve the rolling average bitrate of the modem trace backend, measured over the period defined by the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_PERIOD_MS` Kconfig option.
+To enable logging of the modem trace backend bitrate, set the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_LOG` Kconfig option to ``y``. These logs happen at an interval defined by the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_LOG_PERIOD_MS` Kconfig option.
+If the value of the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_PERIOD_MS` Kconfig option is very low and the value of the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_LOG_PERIOD_MS` Kconfig option is very high, it is sometimes possible to observe highly varying measurements due to the short period over which the rolling average is calculated.
```
```suggestion
To enable logging of the modem trace backend bitrate, set the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_LOG` Kconfig option to ``y``. 
The logging happen at an interval set by the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_LOG_PERIOD_MS` Kconfig option.
If the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_PERIOD_MS` Kconfig option is set to a much lower value than the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_BITRATE_LOG_PERIOD_MS` Kconfig option, you will observe a high variation in measurements due to the short period over which the rolling average is calculated.
```
